Add tracing capabilities to Express

frankrousseau opened the pull request:
Hi Express team,

To follow the discussion posted in the #2191 issue, I send you a proposal to implement tracing support in Express. Let me know if it matches your requirements. It is inspired by Koa Trace.

**Goal**

We want to allow to trace the response life-cycle. The tracing should support any tracing technology. 

**Implementation**
- Add a `tracers` array at the application level.
- Add an `instrument` function on the application prototype to add a tracer to the tracer list.
- Add a `dispatch` function on the application prototype to activate all set tracers.
- Add a `trace` on the response prototype to run tracers with current context and given event as arguments. 
- Add tracing tests.
- Add tracing example.

frankrousseau committed:
tests: add trace tests

* app.instrument
* app._callTracers
* res.trace

frankrousseau committed:
add trace capabilities to express

* allow to add tracer to the app level
* activate tracer from the response object

frankrousseau committed:
examples: add trace example

dougwilson commented on examples/trace/index.js:
```diff
@@ -0,0 +1,42 @@
+  var start = Date.now()
+  res.id = random();
+  res.trace('start', new Date());
+  next();
```
This is not Koa; as soon as you introduce something async in the line below, `res.trace('finish')` not not be the finish. I know it works because everything you have in here is sync, but being an example file, this will confuse people to no end.

frankrousseau replied:
Yep, sorry. I copy/pasted the example. I didn't even think about asynchronism since trace is synchronous.

dougwilson commented on examples/trace/index.js:
```diff
@@ -0,0 +1,42 @@
+
+app.use(function (req, res, next) {
+  res.trace('wait:before');
+  setTimeout(next, Math.random() * 100);
```
what?? Again, not Koa; `wait:after` is going to happen instantly after `wait:before` and both happen before the timer fires.

frankrousseau replied:
... I focused on the code, not on the example.

dougwilson commented on lib/application.js:
```diff
@@ -559,6 +560,43 @@ app.listen = function(){
+ * @api private
+ */
+
+app.dispatch = function(res, event, args){
```
This should probably have a different name, otherwise people will try to `app.dispatch(req, res)` with it.

frankrousseau replied:
Ok, I'll think about a new one. Any suggestion is welcome.

frankrousseau commented:
Thank you for your feedback. I updated the example. Do you think it's ok now?

About the name for the dispatch function. What do you think about `trace` or `runTracers`?

dougwilson commented:
> What do you think about trace or runTracers?

I would probably call it `_callTracers` or something, as it's a private thing to just call the list, right?

frankrousseau commented:
Yes that sounds good. I will commit changes soon.

frankrousseau commented:
`_callTracers` is called by the response object. Should we still mark it as private function? Wouldn't it better to name it `callTracers` instead?

dougwilson commented:
If there is no underscore, we need to document it as if it was a public API that users shoulld be caling themselves. That's the test to know which it should be.

frankrousseau commented:
Ok, I changed the name from `dispatch`  to `_callTracers`.

ilanbiala commented:
@dougwilson are the fixes good?

dougwilson commented:
I want to merge this. Does anyone here have any further comments on this?

raymondfeng commented on lib/application.js:
```diff
@@ -559,6 +560,43 @@ app.listen = function(){
+app._callTracers = function(res, event, args){
+  var date = new Date();
+  var app = this;
+  this.tracers.forEach(function(tracer){
```
I suggest that we use for loop to avoid the performance penalty of closures. See http://jsperf.com/array-foreach-vs-for-loop.

dougwilson replied:
Good idea :) This will get called a lot, so performance is pretty important for this function.

dougwilson commented:
In general, this is pretty much `koa-trace` ported to Express core. @jonathanong do you have any thoughts about the interface (since you've been looking at the interface within `koa`, I'm sure :) ).

jonathanong commented:
https://github.com/strongloop/express/pull/2437/files#diff-65b72b264c47d33fec6edef30f143501R40 way too many arguments here. at this point, might be easier to pass everything through an object. playing with `koa-trace`, i got a little annoyed with the ordering of the arguments `event, date, args` because they were never the order i wanted. if i redid it over again, i would pass an object too.

other than that, it's pretty straight forward and siumple

dougwilson commented:
Cool. Great feedback :)!

frankrousseau commented:
Thank your @jonathanong and @raymondfeng for the feedback. So, if I understand well, what need to be changed is:
- Change forEach function to a tradifional for loop.
- Change `instrument`function arguments to a single object named `options` made of six fields: `app`, `req`, `res`,  `event`, `date` and `args`

Could you me tell if I'm right? Then, if it's ok, I will commit the changes.

dougwilson commented:
Sounds good, yes.

frankrousseau committed:
improve _callTracers function performances

_callTrackers runs a for loop on the application tracer list. Because tracing
should impact performances as least as possible, this commit optimizes this for
loop.

Opitimization performed:

* Usage of traditional for loop instead of forEach
* Caching of tracer array length that is used to test the end of the loop.

Here are the benchmarks that inspired these optimizations:

* http://jsperf.com/array-foreach-vs-for-loop
* http://jsperf.com/for-vs-foreacay

frankrousseau committed:
change call arguments of instrument tracer.

Tracers instrumented at application level now handles a single
option object as argument.

frankrousseau commented:
I made all changes you asked. @dougwilson @jonathanong @raymondfeng it's ready for another review.

jonathanong commented:
i added some comments, but those are usually my personal code styles so i don't think they're a big of a deal. if anything, @dougwilson likes to touch up PRs before he merges them. but overall, i think it's simple and good. i haven't touched express in a while so my opinions matter less ;)

jonathanong commented on examples/trace/index.js:
```diff
@@ -0,0 +1,47 @@
+var debug = require('debug')('trace-example');
+var express = require('../../');
+
+
```
extra space =P

frankrousseau replied:
Thank you for your code style suggestion. But what do you mean by extra space?

jonathanong replied:
There's an extra newline

jonathanong commented on examples/trace/index.js:
```diff
@@ -0,0 +1,47 @@
+var express = require('../../');
+
+
+function random() {
```
i'd move util functions like this to the end of the file.

frankrousseau replied:
Definitely a matter of taste :). But I don't want to break the Express style, so I'll change it.

jonathanong commented on lib/application.js:
```diff
@@ -559,6 +560,64 @@ app.listen = function(){
+
+app.instrument = function(tracer){
+  if (tracer === undefined || typeof tracer  !== 'function')
+    throw new Error('instrument expects a function');
```
should be a `TypeError`

jonathanong commented on lib/application.js:
```diff
@@ -559,6 +560,64 @@ app.listen = function(){
+ */
+
+app.instrument = function(tracer){
+  if (tracer === undefined || typeof tracer  !== 'function')
```
extra space :)

frankrousseau replied:
Same here, what do you mean by extra space? A new blank line?

jonathanong replied:
You have a double space after `tracer`

jonathanong commented on lib/application.js:
```diff
@@ -559,6 +560,64 @@ app.listen = function(){
+  };
+
+  for (var i = 0; i < length; i++) {
+    var tracer = tracers[i];
```
i wouldn't even bother with the `var tracer` statement here but whatever :P

frankrousseau replied:
Yep, I hesitated to put it there.

jonathanong commented on test/app.calltracers.js:
```diff
@@ -0,0 +1,36 @@
+        options.res.id.should.be.equal('1');
+        options.event.should.be.equal('new:event');
+        options.date.should.exist;
+        (options.args[0] === 'arg1').should.be.ok
```
missing semicolon ;)

frankrousseau commented:
Thank you @jonathanong for the additional review. I expect to fix the style soon.

dougwilson commented:
If the only concerns here are style, then updating the PR is optional.

frankrousseau committed:
code style improvement for tracer module

frankrousseau commented:
I fixed the code style. From my point of view, everything is ok (if the travis tests stays green of course).

dougwilson commented on lib/application.js:
```diff
@@ -559,6 +560,63 @@ app.listen = function(){
+ */
+
+app._callTracers = function(res, event, args){
+  var date = new Date();
```
Do we really need a date? Would there ever be such a thing as a delayed tracing call?

frankrousseau replied:
I thought that people who wanted to trace something would be interested to automatically have date for their traces. This line generates it for them. 

But if you consider that it is the tracer's job, it can be removed.

dougwilson replied:
I was merely wondering what the purpose was, and I think you answered it well enough.

dougwilson commented:
Should tracing handle sub apps? i.e. should each app have their own tracing, or if you mount an app as a sub app, should the parent get the events of the sub apps?

jonathanong commented:
@dougwilson i think so

frankrousseau commented:
Do sub apps have a link to their parent app?

dougwilson commented:
> Do sub apps have a link to their parent app?

No, because it's not a relation ship where an app would only have one parent; an app can have multiple parents. You can hook into the `mount` event to setup the links, as seen here: https://github.com/strongloop/express/blob/4.12.0/lib/application.js#L85

frankrousseau commented:
I'm working on it. I have a two questions:
- The expected behavior is: when the child app is reached and fires its tracers, the parent app tracers should be fired too. Is it ok?
- Let's say a child app is mounted on app1 and app2. When a request is sent to app1 and makes the child app run its tracers. Do you expect it runs app1 tracers or both app1 and app2 tracers?

dougwilson commented:
Both of those sound right.

frankrousseau commented:
You can see a proposal in my last two commits. Does it work for you?

frankrousseau committed:
test for parent tracer firing

frankrousseau committed:
link subapp tracer to parent tracer

The operation is performed when the subapp is mounted on the parent app.
A tracer that calls the parent tracers is added to the sub app.

frankrousseau commented:
Hello @dougwilson,

Did you decide anything about this one?

jribeiro commented:
Hi guys, any update on this? Would be really interesting to see this in express.

frankrousseau commented:
I have some time to fix conflicts of this PR. Is tracing still something wanted?

ilanbiala commented:
@dougwilson @expressjs any thoughts on @frankrousseau's comment?

dougwilson commented:
Yes

frankrousseau commented:
Ok, there are not many conflicts. Do you prefer I perform a rebase or a direct merge?

dougwilson commented:
HI @frankrousseau, really whatever is easiest for you. When merged, this would be done as a squash merge, so either way makes no difference. That said, if they are equally easy for you, a rebase makes for an easier to read PR.

frankrousseau commented:
Ok, I did the rebase and there is no conflict anymore with the master branch.

frankrousseau commented:
What is the next step for this PR? @ilanbiala @dougwilson @expressjs

crandmck commented:
@hacksparrow @dougwilson Can we merge this please?

dougwilson commented:
Sorry, didn't notice all these changes to all the feedback. I'm not sure if anyone has actually take a look at all the changes yet (at least, I haven't).

dougwilson commented on test/res.trace.js:
```diff
@@ -0,0 +1,95 @@
+          parentTraces1[0].should.be.equal('one:event:parent1');
+          parentTraces2[0].should.be.equal('one:event:parent2');
+          parentTraces1[1].should.be.equal('one:event:parent1');
+          parentTraces2[1].should.be.equal('one:event:parent2');
```
At some point prior to merging, we'll want to change all these should.js usages to assert.

dougwilson commented on test/res.trace.js:
```diff
@@ -0,0 +1,95 @@
+
+        request(appParent2)
+          .get('/childof2')
+          .expect(200, function(){
```
This `.expect(200` doesn't do anything if you're just ignoring the first argument to the function, which is `err` from the expect.

dougwilson commented on test/res.trace.js:
```diff
@@ -0,0 +1,95 @@
+
+      request(appParent1)
+        .get('/childof1')
+        .expect(200, function(){
```
This `.expect(200` doesn't do anything if you're just ignoring the first argument to the function, which is `err` from the expect.

dougwilson commented on lib/application.js:
```diff
@@ -617,6 +622,63 @@ app.listen = function listen() {
+  };
+
+  for (var i = 0; i < length; i++) {
+    tracers[i](options);
```
Do we want to do any kind of special handling if one of these tracers throws an error, even if that is just to continue to execute all the tracers?

frankrousseau replied:
Good point but I don't know what is best. Do you know what is the best practice in this case?

dougwilson replied:
> Do you know what is the best practice in this case?

I'm not sure, because that is a question that has to do with what exactly the usage of `app.instrument` is supposed to be. Something needs to be done, and I was hoping you would know, since you made the PR :) I can help decide, but need to understand a lot better how this PR would be used in practice, etc. You know, the code here is being built for someone to use, and those users' needs are what this needs to cater towards. I don't really have the need, so it's hard for me to answer this question unless you can help educate me on the needs :)

frankrousseau replied:
I made that PR two years ago based on a open issue. At that time, everyone looked to be fine with the current implementation. So I propose we don't change anything for the moment.

frankrousseau replied:
[Koa version](https://github.com/koajs/trace#koa-trace)
[Related issue](https://github.com/expressjs/express/issues/2191)

dougwilson replied:
If Koa is doing this using a module, why not do that with Express? For example, make a module that works just like the Koa one:

``` js
var app = express()
require('express-trace')(app)
```

dougwilson commented on test/app.calltracers.js:
```diff
@@ -0,0 +1,36 @@
+var express = require('../');
+
+describe('app', function(){
+  describe('._callTracers(options)', function(){
```
I wouldn't bother adding a test for an underscore-prefixed method, unless we're intending it to be a public interface.

dougwilson commented on lib/response.js:
```diff
@@ -1052,6 +1052,32 @@ function sendfile(res, file, options, callback) {
+    default:
+      args = [].slice.call(arguments, 1);
+  }
+  this.app._callTracers(this, event, args);
```
Would there be any benefit to scheduling the tracers run to a new call stack, rather than in the middle of the request trying to be measured?

frankrousseau replied:
Sorry, I don't get it. Can you provide more details (and examples eventually)?

dougwilson replied:
Hm, I'm not sure how to elaborate since I don't fully understand the purpose of the tracers. Can you help me understand what kinds of things people will be putting the tracers? The example provided in this PR shows only a single tracer simply doing `debug(options.res.id, options.event, options.args.join(' '));`. Is that the only thing people will do, or are they expected to add more complex tasks there?

No9 replied:
FWIW The main use case for me is tracing request times as given in the example https://github.com/expressjs/express/pull/2437/files#diff-65b72b264c47d33fec6edef30f143501R8 
So it's quite important that the times measured reflect the actual time taken.

dougwilson replied:
Hi @No9 if that is the use-case, why do we need to have multiple tracers? Can you elaborate on the measurement aspect? I see you pointed to the line `res.start = Date.now()`, but that is in the middleware, so I'm not sure how that would be affected?

No9 replied:
Hey @dougwilson I agree I don't see the need for multiple tracers as multiple destinations are usually handled by the framework that is being used in the instrument call. i.e. I would use something like bunyan as a destination and configure that. 

Apologies I supplied the wrong link https://github.com/expressjs/express/pull/2437/files#diff-65b72b264c47d33fec6edef30f143501R10 the line 2 down from that is what I should have put in.

The use-case I am thinking of is to generate chrome trace output https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit and be able to view it in chrome://tracing tools.

dougwilson replied:
Hi @No9 ah, I see the new line. But the `new Date()` happens in the controller, so I'm still not clear how changing the tracers to run under a different stack would affect their timing accuracy?

No9 replied:
Hey @dougwilson I am not quite getting what you mean by 

> run under a different stack

Is there a snippet you can point to?
If you don't think it will impact the recording of the timestamp but will improve performance then as an end user it makes sense.

dougwilson replied:
@No9 `process.nextTick` / `setImmediate` or similar. The choice between which one of those two or either highly depends on what is expected to happen in the tracing functions the users will set, which is still unclear to me, so if you can provide some good descriptions for all the different use-cases people will do with this, that would help a lot, especially because we need to understand them in order to support the users with this new feature and to get it documented on expressjs.com :)

No9 replied:
@dougwilson Ah yes I see what you mean - and I agree something like `process.nextTick(tracers[i], options);` @ https://github.com/expressjs/express/pull/2437/files/786785b77061bffff90c7b710475af24a35e7ff5#diff-5372f626ee15242f1e2c6eb31655b4faR678 
As the event info has already been recorded your point makes sense.  Can you see any problems @frankrousseau ? 

In terms of use cases the main one (Recording request times) is the main driver and the detail is discussed here 
https://github.com/expressjs/express/issues/2191#issuecomment-62252917 
Providing an abstraction where people can bring there own providers which was the main objection raised in that thread and I think this requests covers it off. 
Another upside of this approach it that developing with and using third party middleware will improve as people will be able to depend on res.trace being there which will improve debugging/trouble shooting. 
My experience with 3rd Party middleware is that people either provide no trace info or bring whatever flavor of logger they see fit and that can be troublesome in terms of integration.

In terms of docs I would be more than happy to flesh out those use cases with @crandmck and support anyway I can.

dougwilson replied:
Ah yes, @No9 that reminds me about something talked about in a recent Express TC meeting: docs. Normally on all repos, adding documentation is a requirement to merge a PR, but with this repo, where the docs are not in the same repo, it's not easy to do that. A suggest was that we should actually make having a PR in https://github.com/expressjs/expressjs.com a blocker to get something merged here, so we know there are actually docs ready to roll when we release the feature, rather than just a promise to maybe make them in the future :) @frankrousseau would need to do this most likely, as otherwise keeping the two PRs in sync between two different people will probably be... hard.

>  My experience with 3rd Party middleware is that people either provide no trace info or bring whatever flavor of logger they see fit and that can be troublesome in terms of integration.

That's a tricky thing this PR does not solve. For example, this will not allow any of the expressjs org middleware to use it, because they are all made to be completely self-contained, no work on anything that provides a Node.js core `req`/`res` object without depending on things attached from in Express itself. If that is a goal, we'll probably need to discuss the best way to accomplish this at the next Express TC meeting and circle back around here to provide some feedback towards that goal.

No9 replied:
Thanks for all this feedback @dougwilson 
I understand the perspective of not hanging express specific features off `req`/ `res`  that could cause incompatibility across the eco-system or if it does make sense then how would that be managed? Either way I hope something lands that answers the initial ask of tracing response times mentioned in #2191 
Look forward to hearing the outcomes from the TC

dougwilson replied:
Hi @No9 wouldn't the use-case outlined at the top of #2191 be better handled with the router events being worked on in PR https://github.com/pillarjs/router/pull/49 ?

frankrousseau replied:
@No9 Could you try using my branch to see if the current state answers to your needs?

No9 replied:
@dougwilson  Oh that is interesting and it looks exactly like what I am looking for thanks - I will have a look during the week.
I also see overall tracing in node is being discussed at the moment https://github.com/nodejs/node/pull/9304

frankrousseau replied:
There are several discussions in this thread, let's put numbers on it to make things easier to follow:

**1) Multiple tracers**

I think we should keep it that way. People may need different levels of tracing and it doesn't make the code much more complex.

**2) processNextTick**

Not sure to understand the need of `processNextTick` here. I suppose that people expect an immediate execution of the tracing and `processNextTick` could confuse them. Can you give me an example? 

**3) Tracing middlewares**

Tracing is very specific to developed application. In which case 3rd party middlewares trace something?

**5) Documentation** 

I will care of it. I will see what could be done for the following use cases: 
- Basic debug calls
- Response time measurement
- Generate Chrome Trace outputs

If I can't find time to do it, I will notify you.

crandmck replied:
Re docs @frankrousseau I created a tracking issue a while back: https://github.com/expressjs/expressjs.com/issues/729.  If there questions or issues specific to how/what needs docs, please put them there.  Thanks!

dougwilson commented on lib/response.js:
```diff
@@ -1052,6 +1052,32 @@ function sendfile(res, file, options, callback) {
+ */
+
+res.trace = function(event) {
+  var args = [];
```
`args` seems to always be reassigned below; probably a useless empty array creation here.

dougwilson commented on test/res.trace.js:
```diff
@@ -0,0 +1,95 @@
+      });
+
+      app.get('/', function(req, res){
+        res.trace('one:event', 'info');
```
In this test, isn't `info` also emitted? I don't see anything verifying that, so we should either verify or just remove the argument here.

frankrousseau replied:
Info parameter is tested here: https://github.com/frankrousseau/express/blob/786785b77061bffff90c7b710475af24a35e7ff5/test/res.trace.js#L15

dougwilson replied:
That link is going to a different test than the one I'm commenting on here.

frankrousseau commented:
Ok @dougwilson, I will have a look at your comments this weekend.

frankrousseau committed:
Proper usage of supertest expect in tracing tests

Supertest lib provides a method called expect to test the result of a request.
This method requires a callback that takes an error object as first parametr.
If this error object is not inspected, the validation performed by expect are
not checked. Tracing tests didn't do that checking.

This commit add checking to trace tests to ensure that no error is returned
when expect method is called.

frankrousseau committed:
In tracing tests, replace should by assert

frankrousseau committed:
In trace function, do not init args to empty array

frankrousseau committed:
Remove _callTracers tests

Private methods do not need to be tested

frankrousseau commented:
I opened a PR for the documentation. It may help to discuss about the use cases. I would be glad to know what you think about it.

https://github.com/expressjs/expressjs.com/issues/733

dougwilson commented:
In case you missed my comment above to one your your responses, I asked why this couldn't be a module like how Koa does it and needs to be built into the core instead. A module could be used against old Express versions, while core cannot do that, for example.

hacksparrow commented:
I am also in favor of making it a "requireable" module, which not only works with Express but also  anything that has a similar signature as Express'. That is unless there is an absolute necessity to bake it into the core.

frankrousseau commented:
Sorry, I missed your question. So, why did I chose to add it to internals? Simply because you, @dougwilson, asked me to do it: https://github.com/expressjs/express/issues/2191#issuecomment-62206562 .

It's now two years we are working on this. So, as a contributor, I would like to have a clear and strong decision from the maintainers: closing this PR or merging it quickly. Our back and forth won't be productive if we keep hesitating on what to do.

dougwilson commented:
> Sorry, I missed your question. So, why did I chose to add it to internals? Simply because you, @dougwilson, asked me to do it: #2191 (comment) .

Hi @frankrousseau the comment you linked to has nothing to do with this PR. This PR is something that can be done in user-space. My comment is about the actual contents of that issue, as explained at the top in https://github.com/expressjs/express/issues/2191#issue-36232871 I don't see anything like that in here, and to add those events, it would have to be done in core, but this PR has nothing to do with adding any events at all.

I'm sorry if there was miscommunication around that, which is probably what is causing this PR to take so long. What can we do to get some clear communication going between us? Would it be effective to have a voice / video meeting between us over the weekend to hash this out or something?

frankrousseau commented:
This comment is definetely related with this PR. If you read the following comments, you agreed with the current implementation. Moreover, this comment match with my explicit question (which is right above): 

>  should I make a separate module like the koa one or should I contribute directly to the Express platform?

I can understand you changed your mind in 2 years, but now we need to find a consensus about what what is expected and if that feature is still wanted or not. Having a chat this weekend sounds a good solution. It will be also the opportunity for you to explain me the contribution lifecycle for the ExpressJs project. When are you available? What is your timezone mine is Europe/Paris, UTC+01:00)?

dougwilson commented:
I think the resolution to this PR is pretty much settled: make a module on npm.

frankrousseau commented:
Thank you for making things clearer.

dougwilson commented:
Hi @frankrousseau I still do want to have the discussion, if you're up to it. You never know what will come from it (even a re-open). I am in US Eastern time (UTC-5 right now) and so something like 10am or later is workable. If you ping me an email, we can work out the contact details and time. We could even make it a Google Hangout on our YouTube channel if you're up to that, so everyone can listen it / have a record :D !

dougwilson commented:
I know there has been a lot of miscommunication all around, and I am sorry. I feel bad for leading around just as much for you feeling bad for hanging around, I'm sure. FWIW, now with a much better understanding of what this PR actually does, and how it is implemented, I think that since it is capable to be implemented in user-land, actually implementing is user land would have a better positive impact than constraining it to core as @hacksparrow noted above. There is also no reason why a future core can't, for example, just automatically include the user-land mode as the method to add to core by default, similar to how much of core currently works.

frankrousseau commented:
Now I know better Express I understand your point. 

But what frustrates me is that I was explicit about what I was going to do, I integrated many feedbacks from numerous reviews. To finally read, 2 years later, that it should be done outside of Express. While the module separation was one of my first question. To avoid this kind of situation, I think that something could be changed in the way contributions are handled. 

I will send you an email to set an Hangout meeting or equivalent.

dougwilson commented:
Hi @frankrousseau I understand the frustration. Also keep in mind that over those two years, the express repo has been moved between organization, has changed ownership, and more. The project was not in a good position to make large contributions successful and finally, after all this time, we are getting a Technical Committee to oversee Express, getting documentation around the principals of Express, and more. The only good thing I can say about your situation is that the contribution was over the duration of a very turbulent time in Express's history and you have been caught in that situation.

frankrousseau commented:
Ok, so if things are better structurated, that's a good news. I hope things will keep improving. Making a clear choice about this PR is already a sign of progress.

frankrousseau commented:
I created a separated module named `express-tracer`. We can do fast iteration on it. I published a first version, but it requires feedback from real life use cases to be fully operational. So, @No9 @crandmck @jribeiro, I would be glad to know what you think about it :).

https://github.com/frankrousseau/express-tracer

crandmck commented:
I have some minor comments on the README...
  
@dougwilson Should we fork (or copy?) this in the expressjs org before we start working on it?  Or, what's the proper workflow....?

No9 commented:
@frankrousseau  thanks 
I'll take it for a spin with my dtrace implementation and feed back anything in the new repo
